### PR TITLE
Skill specific tasks #6

### PR DESCRIPTION
**Pull request by rmd6502**
Not yet working, just creating this so people can take a look and hopefully unstick me
Problem I'm having is I can't get the categories to display in the Task edit page.  If I uncomment the commented-out code in task.rb they display, but the page is extremely ugly and the selected categories aren't written to the categories_tasks many:many join table.

Edit: This is now working.

### CODE REVIEW

**Comment by rmd6502** [comment]
Only one test case still fails - reviewing the semantics of left_outer_join to figure out why

**Comment by rmd6502** [comment]
All test cases pass, awaiting instructions!

File: Gemfile [hunk 1]
```diff
@@ -2,7 +2,7 @@ source 'https://rubygems.org'
 
 
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
-gem 'rails', '~> 5.0.0'
+gem 'rails', '~> 5.0.1'
```
Comment (MiguelCarranza) [inline Gemfile:5]: any particular reason for upgrading Rails?

File: db/migrate/20170121203621_create_categories.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class CreateCategories < ActiveRecord::Migration[5.0]
+  def change
+    create_table :categories do |t|
+      t.string :name
```
Comment (MiguelCarranza) [inline db/migrate/20170121203621_create_categories.rb:4]: maybe index this guy, but I guess this table is not going to be huge

Comment (MiguelCarranza) [reply db/migrate/20170121203621_create_categories.rb:4]: actually, yeah, make it a non-unique index

File: app/admin/task.rb [hunk 3]
```diff
@@ -1,6 +1,48 @@
...
+  index do
+    id_column
+    column :description
+    column :is_multi_use
+    column :has_expired
+    column :bounty
+    column :categories do |t|
+      t.categories.map do |c|
+        c.name
+      end.join(', ')
```
Comment (MiguelCarranza) [inline app/admin/task.rb]: Maybe it's me, but the `end.join` got me like 😕 
maybe it's clearer with a one liner, like ` t.categories.map { |c| c.name }.join(', ')`

File: app/models/user.rb [hunk 4]
```diff
@@ -1,12 +1,42 @@
...
+    else
+      # tasks that have categories
+      category_task_ids = Task.joins(:categories).distinct.pluck(:id)
+      # tasks that match the user's categories
+      candidate_tasks = Task.where(has_expired: false)
+                           .where.not(id: excluded_task_ids)
+                           .left_outer_joins(:categories)
+                             .where(categories: { id: user_skills })
+      # tasks that don't have any categories
+      tasks_without_categories = Task.where(has_expired: false)
```
Comment (MiguelCarranza) [inline app/models/user.rb]: I am little scared of the `has_expired` thing being repeated in multiple times of the code now. Maybe we should create something like `unexpired_tasks` private method or something

**Comment by MiguelCarranza** [comment]
@rmd6502 besides my small comments, looking great! thanks a lot :)

File: Gemfile (hunk 1 above)
Comment (rmd6502) [reply Gemfile:5]: Yeah, was getting a weird bug in ActiveRecord - I'll check specifics if you want, don't remember offhand.  Stackoverflow said the bug was fixed in 5.0.1

File: db/migrate/20170121203621_create_categories.rb (hunk 2 above)
Comment (rmd6502) [reply db/migrate/20170121203621_create_categories.rb:4]: done

**Comment by rmd6502** [comment]
Thanks for the review!  My intuition tells me that there should be a better way than the map call to make a comma separated list of categories, but my ruby is 4 years out of date, not counting last weekend.

File: Gemfile (hunk 1 above)
Comment (MiguelCarranza) [reply Gemfile:5]: 👍

**Comment by MiguelCarranza** [comment]
awesome, merging and deploying!

**Review by MiguelCarranza — APPROVED** [review]

**Pull request merged**